src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -520,12 +524,18 @@ public void AddLocalRpcMethod(string rpcMethodName, MethodInfo handler, object t
         /// </summary>
         public void StartListening()
         {
-            this.startedListening = true;
-
             Verify.Operation(this.MessageHandler.CanRead, Resources.StreamMustBeReadable);
             Verify.Operation(this.readLinesTask == null, Resources.InvalidAfterListenHasStarted);
             Verify.NotDisposed(this);
-            this.readLinesTask = Task.Run(this.ReadAndHandleRequestsAsync, this.disposeCts.Token);
+
+            // We take a lock around this Task.Run and field assignment,
+            // and also immediately within the invoked Task itself,
+            // to guarantee that the assignment will complete BEFORE we actually read the first message.
+            // See the StartListening_ShouldNotAllowIncomingMessageToRaceWithInvokeAsync test.
+            lock (this.syncObject)
+            {
+                this.readLinesTask = Task.Run(this.ReadAndHandleRequestsAsync, this.disposeCts.Token);
```

Comment:
just to make sure Task.Run run things in other thread, put TaskScheduler.Default explicitly?

Reply:
ah. ignore what I wrote. looks like Task.Run always use TaskScheduler.Default and you can't change it.